iopipe/contrib/trace/auto_http.py

```diff
@@ -0,0 +1,137 @@
+import functools
+import uuid
+
+try:
+    from requests.sessions import Session
+except ImportError:
+    Session = None
+
+try:
+    from botocore.vendored.requests.sessions import Session as BotocoreSession
+except ImportError:
+    BotocoreSession = None
+
+from iopipe.compat import urlparse
+
+if Session is not None:
+    original_session_send = Session.send
+
+if BotocoreSession is not None:
+    original_botocore_session_send = BotocoreSession.send
+
+EXCLUDE_HEADERS = ['Authorization', 'Cookie', 'Proxy-Authorization', 'Set-Cookie']
+
+REQUEST_KEYS = [
+    ('method', 'method'),
+    ('url', 'url'),
+]
+
+RESPONSE_KEYS = [
+    ('status_code', 'statusCode'),
+]
+
+URLPARSE_KEYS = [
+    ('fragment', 'hash'),
+    ('hostname', 'hostname'),
+    ('path', 'path'),
+    ('port', 'port'),
+    ('query', 'query'),
+    ('scheme', 'protocol'),
+]
+
+
+def patch_session_send(context, auto_measure, http_filter):
+    """
+    Monkey patches requests' Session class, if available. Overloads the
+    send method to add tracing and metrics collection.
+    """
+    if Session is None:
+        return
+
+    def send(self, *args, **kwargs):
+        id = str(uuid.uuid4())
+        with context.iopipe.mark(id):
```

Comment:
How do we want to name auto-traces? UUID seems especially not-descriptive

Reply:
The user doesn't need to be concerned with the UUID, we use it to associate traces with their respective HTTP request. In the dashboard, we'll likely hide the UUID completely and just show the HTTP request and the relevant trace.

Reply:
How do we distinguish between a UUID we generate and what a user creates? IMO this should be something meaningful, otherwise showing traces without associating with any concept or context won't have meaning

Reply:
I accidentally commented in the wrong context, but I don't think these should be UUIDs. They should have some meaningful context, otherwise they are measurements without meaning.

Reply:
We won't be displaying a trace as a UUID. Instead, we map the UUID to the relevant HTTP request and display information about that HTTP request with the trace.